File: leetcode/rotate_list.js
// https://leetcode.com/problems/rotate-list/
// 61. Rotate List
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function (head, k) {
    let count = 0;
    let tail = head;
    while (tail) {
        count++;
        if (!tail.next) {
            break;
        }
        tail = tail.next;

    }
    if (!count) {
        return head
    }
    let pointer = head;
    tail.next = head;
    let splitByIndex = count - k % count;
    while (splitByIndex > 1) {
        pointer = pointer.next;
        splitByIndex--;
    }
    let newHead = pointer.next;
    pointer.next = null;
    return newHead;
};

function main() {
    let array = [1, 2, 3, 4, 5];
    let k = 8
    let head = getLinkedListFromArray(array);
    let res = rotateRight(head, k);
    let resArray = getArrayFromList(res);
    console.log(resArray);
}

main();

function getArrayFromList(head) {
    let pointer = head;
    let res = [];
    while (pointer) {
        res.push(pointer.val)
        pointer = pointer.next;
    }
    return res;
}

function getLinkedListFromArray(arr) {
    if (!arr.length) {
        return null
    }
    let head = new ListNode();
    let pointer = head;
    for (let i = 0; i < arr.length; i++) {
        pointer.val = arr[i];
        if (i === arr.length - 1) {
            pointer.next = null
        } else {
            pointer.next = new ListNode();
            pointer = pointer.next
        }

    }
    return head;
}

function ListNode(val, next) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
}

File: leetcode/rotate_list.py
from typing import Optional, List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def print(self):
        arr = []
        head = self
        while head is not None:
            arr.append(head.val)
            head = head.next
        print(arr)


def constructLinkedList(list: List[int]) -> Optional[ListNode]:
    ll = ListNode(list[0])
    head = ll
    for index in range(1, len(list)):
        node = ListNode(list[index], None)
        head.next = node
        head = node
    return ll


class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        total = 0
        start = head
        p1 = head
        while start is not None:
            total += 1
            start = start.next
        if total == 0:
            return head
        k = k % total
        if k == total or k == 0:
            return head
        move = total - k
        while move > 1:
            p1 = p1.next
            move = move - 1
        p2 = p1
        newHead = p1.next
        while p2.next is not None:
            p2 = p2.next

        p2.next = head
        p1.next = None
        return newHead


if __name__ == '__main__':
    l = [1, 2, 3, 4]
    k = 2
    ll = constructLinkedList(l)
    res = Solution().rotateRight(ll, k)
    res.print()
